### PR TITLE
New Analytics Method : Coreness

### PR DESCRIPTION
**Pull request by duckofyork**
# Why is this PR needed?
New analytics method: centrality based on k-core, or "coreness"
# What does the PR do?
Add coreness query and its impl
# Does it break backwards compatibility?
No
# List of future improvements not on this PR
Unify degree and coreness, as they are both centrality measures.
Unify connected component and k-core, as they are cluster queries.
Update graql syntax.

### CODE REVIEW

File: grakn-graql/src/main/java/ai/grakn/graql/internal/query/analytics/CorenessQueryImpl.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    }
+
+    @Override
+    public Map<Integer, Set<String>> execute() {
+        LOGGER.info("Coreness query is started");
+        long startTime = System.currentTimeMillis();
+
+        if (k < 2) throw GraqlQueryException.kValueSmallerThanTwo();
+
+        includeAttribute = true; //TODO: REMOVE THIS LINE
```
Comment (aelred) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/query/analytics/CorenessQueryImpl.java]: Can we remove it now?

File: grakn-graql/src/test/java/ai/grakn/graql/internal/analytics/CorenessTest.java [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+    @ClassRule
+    public final static SessionContext sessionContext = SessionContext.create();
+
+    @Before
+    public void setUp() {
+        session = sessionContext.newSession();
+    }
+
+    @Test(expected = GraqlQueryException.class)
+    public void testKSmallerThan2() {
```
Comment (aelred) [inline grakn-graql/src/test/java/ai/grakn/graql/internal/analytics/CorenessTest.java]: We should explain that we expect this to throw in the test name, like `testKSmallerThan2_ThrowsException`

File: grakn-graql/src/test/java/ai/grakn/graql/internal/analytics/CorenessTest.java [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+        try (GraknTx graph = session.open(GraknTxType.READ)) {
+            graph.graql().compute().coreness().minK(1).execute();
+        }
+    }
+
+    @Test
+    public void testOnEmptyGraph() {
```
Comment (aelred) [inline grakn-graql/src/test/java/ai/grakn/graql/internal/analytics/CorenessTest.java]: We can also say the expected result here, like `testOnEmptyGraph_ReturnsEmptyMap`. Same for the tests below too!

File: grakn-graql/src/main/java/ai/grakn/graql/internal/query/analytics/CorenessQueryImpl.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    }
+
+    @Override
+    public CorenessQuery of(String... ofTypeLabels) {
+        if (ofTypeLabels.length > 0) {
+            ofTypeLabelsSet = true;
+            this.ofLabels = Arrays.stream(ofTypeLabels).map(Label::of).collect(Collectors.toSet());
+        }
+        return this;
+    }
```
Comment (aelred) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/query/analytics/CorenessQueryImpl.java]: Even better is if we define this in terms of the other `of` method below!

```
public CorenessQuery of(String... ofTypeLabels) {
    return of(Arrays.stream(ofTypeLabels).map(Label::of).collect(Collectors.toSet());
}
```

File: grakn-graql/src/main/java/ai/grakn/graql/internal/query/analytics/CorenessQueryImpl.java (hunk 1 above)
Comment (duckofyork) [reply grakn-graql/src/main/java/ai/grakn/graql/internal/query/analytics/CorenessQueryImpl.java]: Not yet. I will do it in another pr, so we can remove it from all analytics queries

File: grakn-graql/src/test/java/ai/grakn/graql/internal/analytics/CorenessTest.java [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+
+            entityId1 = entity1.getId();
+            entityId2 = entity2.getId();
+            entityId3 = entity3.getId();
+            entityId4 = entity4.getId();
+
+            graph.commit();
+        }
+    }
+}
```
Comment (kasper-piskorski) [inline grakn-graql/src/test/java/ai/grakn/graql/internal/analytics/CorenessTest.java]: Might be worth describing the configuration graphically, saves time when dealing with test failures.

File: grakn-graql/src/test/java/ai/grakn/graql/internal/analytics/CorenessTest.java [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+            graph.putRelationshipType(veryRelated)
+                    .relates(role3).relates(role4)
+                    .addRelationship()
+                    .addRolePlayer(role3, entity1)
+                    .addRolePlayer(role4, entity2);
+
+            Map<Integer, Set<String>> result = graph.graql().compute().coreness().execute();
+            assertTrue(result.isEmpty());
+        }
+    }
```
Comment (kasper-piskorski) [inline grakn-graql/src/test/java/ai/grakn/graql/internal/analytics/CorenessTest.java]: Just trying to have a general understanding. I understand the default k is 2. So when executing coreness for what values of k are calculations performed? Single specific value (here the default I guess)? Range of values starting from min k?

Comment (duckofyork) [reply grakn-graql/src/test/java/ai/grakn/graql/internal/analytics/CorenessTest.java]: A vertex u has coreness c if it belongs to a c-core but not to any (c+1)-core.
In coreness query, we start with k = minK, keep persisting and updating coreness of vertices.

File: grakn-graql/src/test/java/ai/grakn/graql/internal/analytics/CorenessTest.java (hunk 5 above)
Comment (duckofyork) [reply grakn-graql/src/test/java/ai/grakn/graql/internal/analytics/CorenessTest.java]: That's a good idea man. I will try to draw this in the test, something like this:
```/*
        root    10
    	(0) -------> (1) \
    	 |  \       /  ^  \
    	 |   \30   |   |20 \
    	 |10  \    |10 |    \10
    	 |     \   |  /      \
    	 V  15  V  V /   20   V
    	(3)<----- (2) -----> (4)
    	  \-------^
    	     40
    	 */

**Pull request merged**